Финальное задание спринта 7

zolotyh-dk opened the pull request:
### Добавлена новая реализация менеджера задач `FileBackedTaskManager`

- Новый менеджер задач дополняет функции `InMemoryHistoryManager` возможностью сохранения и восстановления задач из .csv файла
- Добавлен новый тестовый класс `FileBackedTaskManager`, который наследуется от `InMemoryTaskManagerTest` и дополняет его тестами на корректную работу с файлом бэкапа
- Класс `Task` теперь реализует интерфейс `Comparable` для сортировки по ID перед записью задач в файл. Это позволяет при десериализации гарантировать что всем типам задач присвоятся старые ID.

zolotyh-dk committed:
добавить FileBackedTaskManager, изменить toString всех типов задач под формат CSV

zolotyh-dk committed:
Добавить класс Deserializer с методом преобразования строки в задачу

zolotyh-dk committed:
реализовать метод save для сериализации всех задач в файл .csv

zolotyh-dk committed:
добавить исключение ManagerSaveException и его выброс в методе save

zolotyh-dk committed:
добавить метод загрузки из файла loadFromFile, добавить сценарий использования менеджера в FileBackedTaskManager.main()

zolotyh-dk committed:
рефакторинг методов класса Deserializer

zolotyh-dk committed:
добавить класс FileBackedTaskManagerTest

zolotyh-dk committed:
Добавить сортировку задач по id перед сохранением в файл

zolotyh-dk committed:
Добавить тесты для модифицирующих операций в FileBackedTaskManager

zolotyh-dk committed:
Наследовть FileBackedTaskManagerTest от InMemoryTaskManagerTest, дополнить тестами на восстановление задач из файла

zolotyh-dk committed:
Отформатировать код, добавить final где необходимо

zolotyh-dk committed:
Добавить каталог test_resources в коммит

YazykovMikhail commented on src/main/ru/yandex/javacourse/zolotyh/schedule/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,199 @@
+        save();
+    }
+
+    private void save() {
```
В целом ок, но есть пару моментов
1) '\n' - в разных операционных системах перевод строки пишется по разному
2) сотировать не обязательно, генерировать id при считываении из истории не нужно, ведь некоторые задчи могли быть удали и в истории было бы 1,2,5,7 если генерировать id то восстановится не актуальными занчениями
3) Лучше использовать буферизацию для записи, будет экономить ресурсы

Вот так можно улучшить:
```

	protected void save() {
		try (BufferedWriter writer = new BufferedWriter(new FileWriter(file))) {
			writer.write(CSVTaskFormat.getHeader());
			writer.newLine();

			for (Map.Entry<Integer, Task> entry : tasks.entrySet()) {
				final Task task = entry.getValue();
				writer.write(CSVTaskFormat.toString(task));
				writer.newLine();
			}

			for (Map.Entry<Integer, Subtask> entry : subtasks.entrySet()) {
				final Task task = entry.getValue();
				writer.write(CSVTaskFormat.toString(task));
				writer.newLine();
			}

			for (Map.Entry<Integer, Epic> entry : epics.entrySet()) {
				final Task task = entry.getValue();
				writer.write(CSVTaskFormat.toString(task));
				writer.newLine();
			}

			writer.newLine();
		} catch (IOException e) {
			throw new ManagerSaveException("Can't save to file: " + file.getName(), e);
		}
	}
```

zolotyh-dk replied:
Переделал метод `save()` с использованием буферизации, сортировку убрал, '\n' везде по коду заменил на `BufferedWriter.newLine()` и `System.lineSeparator()`

YazykovMikhail commented on src/main/ru/yandex/javacourse/zolotyh/schedule/task/Epic.java:
Стоит все поля в классе сделать protected, что бы был прямой доступ из класса наслденика

YazykovMikhail commented on src/main/ru/yandex/javacourse/zolotyh/schedule/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,199 @@
+
+        for (int i = 1; i < lines.length; i++) {
+            final Task task = fromString(lines[i]);
+            if (task instanceof Epic) {
```
Не зря ты вводил типы задач, используй их

zolotyh-dk replied:
Вынес определение типа и добавление в хранилища в отдельный метод `void addAnyTask(Task task)`

YazykovMikhail commented on src/main/ru/yandex/javacourse/zolotyh/schedule/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,199 @@
+        for (int i = 1; i < lines.length; i++) {
+            final Task task = fromString(lines[i]);
+            if (task instanceof Epic) {
+                taskManager.addNewEpic((Epic) task);
```
Как уже скзаал при сохранении, методы добавления использовать не стоит пр ивосстанволении, потому что методы гененируют новые id, а нам неоходимо именно восстановить текущие. Непосредтсвеннно добавляй сразу в хранилизе считанные задачи

zolotyh-dk replied:
Исправил, для прямого доступа к хранилищам сделал поля protected

YazykovMikhail commented on src/main/ru/yandex/javacourse/zolotyh/schedule/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,199 @@
+                taskManager.addNewTask(task);
+            }
+        }
+
```
И нужно восстанавливать счетчик id, как ты уже понял задач может быть сохранено в файл 10, но последний идентификатор может быть 50, потому что в овремя работы приложения задачи удалялись.

YazykovMikhail commented on src/main/ru/yandex/javacourse/zolotyh/schedule/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,199 @@
+        }
+    }
+
+    public static FileBackedTaskManager loadFromFile(File file) {
```
Сначала прочитай все комментари к методу, 


СПОЙДЕР:
```
public static FileBackedTaskManager loadFromFile(File file) {
		final FileBackedTaskManager taskManager = new FileBackedTaskManager(file);
		try {
			final String csv = Files.readString(file.toPath());
			final String[] lines = csv.split(System.lineSeparator());
			int generatorId = 0;
			for (int i = 1; i < lines.length; i++) {
				String line = lines[i];
				if (line.isEmpty()) {
					break;
				}
				final Task task = CSVTaskFormat.fromString(line);
				final int id = task.getId();
				if (id > generatorId) {
					generatorId = id;
				}
				taskManager.addAnyTask(task);
			}
			for (Map.Entry<Integer, Subtask> e : taskManager.subtasks.entrySet()) {
				final Subtask subtask = e.getValue();
				final Epic epic = taskManager.epics.get(subtask.getEpicId());
				epic.addSubtaskId(subtask.getId());
			}
			taskManager.generatorId = generatorId;
		} catch (IOException e) {
			throw new ManagerSaveException("Can't read form file: " + file.getName(), e);
		}
		return taskManager;
	}
```

YazykovMikhail commented on src/main/ru/yandex/javacourse/zolotyh/schedule/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,199 @@
+        try (FileWriter writer = new FileWriter(backup, StandardCharsets.UTF_8)) {
+            writer.write(heading + '\n');
+            for (int i = 0; i < all.size(); i++) {
+                writer.write(all.get(i).toString());
```
Стоит использовать отдельны метод toString, не зря в задании говорится : 
"Напишите метод сохранения задачи в строку"
Метод toString которые есть в каждом обьекте, он используется внутри самой джавы для многих вещей, а тут конкретно бизнес процесс, не стоит их мешать между собой.

YazykovMikhail replied:
Очень просто метод:
```
public static String toString(Task task) {
		return task.getId() + "," + task.getType() + "," + task.getName() + "," + task.getStatus() + "," + task.getDescription() + "," + (task.getType().equals(TaskType.SUBTASK) ? task.getEpicId() : "");
	}
```

zolotyh-dk replied:
Добавил метод `toString()` в `CSVTaskFormat`.  Предложенный вариант доработал, чтобы запятая в конце строки ставилась и для подзадач тоже + добавил каст до эпика перед `task.getEpicId()`

YazykovMikhail commented on src/main/ru/yandex/javacourse/zolotyh/schedule/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    //Формат строки: "id,type,name,status,description,epicId"
+    private static Task fromString(String value) {
```
Стоит все методы из пункта задания:
"Как сохранять задачи в файл и считывать их из него" вынести в утилитный класс:

CSVTaskFormat и выызвать от туда статические методы

YazykovMikhail commented on src/main/ru/yandex/javacourse/zolotyh/schedule/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    private void save() {
+        final String heading = "id,type,name,status,description,epic";
```
Тоже стоит вынести в утилитный класс CSVTaskFormat
Ввиде константы или метода:
```
	public static String getHeader() {
		return "id,type,name,status,description,epic";
	}

```

zolotyh-dk replied:
Вынес в `CSVTaskFormat`

YazykovMikhail commented on src/main/ru/yandex/javacourse/zolotyh/schedule/task/Task.java:
```diff
@@ -1,10 +1,11 @@
 import java.util.Objects;
 
-public class Task {
+public class Task implements Comparable<Task> {
```
Стоит добавить или поле Type type или в каждую задачу просто один метод:

```
	public TaskType getType() {
		return TaskType.TASK;
	}
```
Который будет возвращать конкретный тип задачи

zolotyh-dk replied:
Добавил методы в `Task` и переопределил в наследниках

YazykovMikhail reviewed (commented):
В класс Manages стоит обновить работу поумолчнию, теперь она будет работать через файл:

	public static TaskManager getDefault() {
		return new FileBackedTaskManager(new File("resources/task.csv"));
	}

YazykovMikhail reviewed (changes requested):
Вообще хорошо, есть небольшие моменты которые стоит поправить

zolotyh-dk committed:
Добавить класс CSVTaskFormat с методом toString(Task)
Вернуть задачам оригинальные toString, добавить задачам методы getType

zolotyh-dk committed:
Отрефакторить метод FileBackedTaskManager.save()
Использовать буферизацию и прямое обращение к хранилищу

zolotyh-dk committed:
Отрефакторить метод loadFromFile(File)

Задачи сохранять напрямую в хранилище, воостанавливать значение generatorId, метод fromString вынести в класс CSVTaskFormat, для переноса строки использовать System.lineSeparator()

zolotyh-dk committed:
В методе fromString добавить десериализацию статуса эпика

- Переделать тесты для работы с System.lineSeparator() вместо \n
- В классе Task убрать имплементацию Comparable, т.к. сортировка не нужна, убрать тест для compareTo
